src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
       segment_meta_cstore::impl,
       serde::version<0>,
       serde::compat_version<0>> {
+    // TODO tunable?
+    constexpr static auto max_buffer_entries = 1024u;
+
 public:
     void append(const segment_meta& meta) { _col.append(meta); }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl> begin() const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.begin());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl> end() const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.end());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     find(model::offset o) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.find(o()));
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     lower_bound(model::offset o) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.lower_bound(o()));
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     upper_bound(model::offset o) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.upper_bound(o()));
     }
 
     std::optional<segment_meta> last_segment() const {
+        if (!_write_buffer.empty()) {
+            auto last_committed_offset = _col.last_committed_offset().value_or(
+              model::offset::min());
+            auto buffer_last_base_offset = _write_buffer.crbegin()->first;
+            // CHECK: base_offset is never equal to committed_offset
+            if (likely(buffer_last_base_offset > last_committed_offset)) {
+                // return last one in _write_buffer
+                return _write_buffer.crbegin()->second;
+            }
+            // fallthrough
+        }
         return _col.last_segment();
     }
 
-    void insert(const segment_meta& m) { _col.append(m); }
+    void insert(const segment_meta& m) {
+        auto [m_it, insertion_took_place] = _write_buffer.insert_or_assign(
```

Comment:
Also, maybe the logic will get a bit more straightforward if you will check for the overlap before insertion. Then, some action could be taken on replaced element inside the `_write_buffer` instead of figuring out what was replaced afterwards.

Reply:
so in the end column_store::insert_entries handles all the buffer directly. this makes the append-only operation still fast, since the whole merging section is skipped. It also buys atomicity, since it's done on list of clones that will be swapped at the end of operation.